common/guestchain/src/config.rs

```diff
@@ -91,3 +91,118 @@ pub struct Config {
     /// epoch.
     pub min_epoch_length: crate::height::HostDelta,
 }
+
+#[derive(Clone, Debug, borsh::BorshSerialize, borsh::BorshDeserialize)]
+pub struct UpdateConfig {
+    pub min_validators: Option<NonZeroU16>,
+    pub max_validators: Option<NonZeroU16>,
+    pub min_validator_stake: Option<NonZeroU128>,
+    pub min_total_stake: Option<NonZeroU128>,
+    pub min_quorum_stake: Option<NonZeroU128>,
+    pub min_block_length: Option<crate::height::HostDelta>,
+    pub max_block_age_ns: Option<u64>,
+    pub min_epoch_length: Option<crate::height::HostDelta>,
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum UpdateConfigError {
+    /// Minimum validators are more than existing
+    ///
+    /// If minimum validators are higher than existing, then the
+    /// none of the existing validators can leave unless the validators are more
+    /// than the minimum.
+    MinValidatorsHigherThanExisting,
+
+    /// Maximum validators should always be equal or higher than minimum validators
+    MaxValidatorsCannotBeLowerThanMin,
+
+    /// Minimum Total Stake is higher than existing
+    ///
+    /// If minimum total stake is higher than existing, then none of them
+    /// can withdraw their unless the total stake is more than the minimum.
+    MinTotalStakeHigherThanExisting,
+
+    /// Minimum total stake is lower than minimum quorum stake
+    ///
+    /// If minimum total stake is lower than minimum quorum stake, then
+    /// the total stake can be less than quorum and block would never be
+    /// finalized since the total stake is less than quroum stake.
+    MinTotalStakeHigherThanMinQuorumStake,
+
+    /// Minimum Quorum Stake is higher than existing total stake
+    ///
+    /// If minimum quorum stake is higher than existing total stake, then
+    /// blocks would never get finalized until more stake is added and quorum
+    /// stake is less than head stake.
+    MinQuorumStakeHigherThanTotalStake,
+}
+
+impl Config {
+    pub fn update(
+        &mut self,
```

Comment:
Actually, it just occurred to me that we need to leave `self`
unchanged on an error.  So inside of the function I think we need to
operate on the copy and commit only once everything is done:

```
    pub fn update(&mut self, ...) -> Result<(), UpdateConfigError> {
        let mut this = self.clone();
        ...
        *self = this
    }
```

Alternatively have an `pub fn updated(self, ...) -> Result<Self, ...>`
method and caller would do `self.config =
self.config.clone().update(...)?;`.

Reply:
isnt it atomic?

Reply:
like if it fails anywhere in between, the changes would be reverted.

Reply:
If you’re talking Solana transaction sure, but this is a type which is blockchain-agnostic and could be used outside of blockchain. Because of that I’d rather it didn’t assume that operation get reverted.

Reply:
ya right, lemme change that then